File: backend/src/main/java/boukevanzon/Anchiano/model/Task.java
package boukevanzon.Anchiano.model;

import boukevanzon.Anchiano.enums.TaskPriority;
import boukevanzon.Anchiano.enums.TaskStatus;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.List;
import java.util.ArrayList;
import java.util.Set;
import java.util.HashSet;

@Entity
@Table(name = "tasks")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Task {
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(optional = false)
    @JoinColumn(name = "workspace_id")
    private Workspace workspace;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(nullable = true)
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TaskStatus status = TaskStatus.TODO;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TaskPriority priority = TaskPriority.MEDIUM;

    @Column(nullable = true)
    private LocalDateTime dueDate;

    // 👇 wijziging: meerdere assignees
    @ManyToMany
    @JoinTable(
        name = "task_assignees",
        joinColumns = @JoinColumn(name = "task_id"),
        inverseJoinColumns = @JoinColumn(name = "user_id")
    )

    private Set<User> assignees = new HashSet<>();

    // 👇 wijziging: labels via element collection
    @ElementCollection
    @CollectionTable(name = "task_labels", joinColumns = @JoinColumn(name = "task_id"))
    @Column(name = "label")
    private List<String> labels = new ArrayList<>();

    @ManyToOne(optional = false)
    @JoinColumn(name = "created_by")
    private User createdBy;
    @Column(nullable = false)
    private LocalDateTime createdAt = LocalDateTime.now();
    @Column(nullable = false)
    private LocalDateTime updatedAt = LocalDateTime.now();
    @Column(nullable = true)
    private LocalDateTime deletedAt;
}

File: frontend/lib/features/tasks/models/task.dart
import '../../auth/models/user.dart';

enum TaskStatus { TODO, IN_PROGRESS, DONE }
enum TaskPriority { LOW, MEDIUM, HIGH }

class Task {
  final int id;
  final int workspaceId;
  final String title;
  final String description;
  final TaskStatus status;
  final TaskPriority priority;
  final DateTime? dueDate;
  final List<String> labels; // simplificatie (later model uitbreiden)
  final List<User> assignees;
  final User createdBy;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? deletedAt;

  Task({
    required this.id,
    required this.workspaceId,
    required this.title,
    required this.description,
    required this.status,
    required this.priority,
    this.dueDate,
    required this.labels,
    required this.assignees,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
    this.deletedAt,
  });

  factory Task.fromJson(Map<String, dynamic> json) {
    TaskStatus parseStatus(String s) =>
        TaskStatus.values.firstWhere((e) => e.name == s);
    TaskPriority parsePriority(String p) =>
        TaskPriority.values.firstWhere((e) => e.name == p);

    return Task(
      id: json['id'],
      workspaceId: json['workspaceId'],
      title: json['title'],
      description: json['description'],
      status: parseStatus(json['status']),
      priority: parsePriority(json['priority']),
      dueDate: json['dueDate'] != null ? DateTime.parse(json['dueDate']) : null,
      labels: (json['labels'] as List<dynamic>?)
              ?.map((l) => l['name'] as String)
              .toList() ??
          [],
      assignees: (json['assignees'] as List<dynamic>?)
              ?.map((a) => User.fromJson(a))
              .toList() ??
          [],
      createdBy: User.fromJson(json['createdBy']),
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
      deletedAt:
          json['deletedAt'] != null ? DateTime.parse(json['deletedAt']) : null,
    );
  }
}
